mod_muc_focus.lua

```diff
@@ -1062,6 +1060,43 @@ module:hook("iq/bare", function (event)
                 :up() -- content
             end
 
+            -- update the participant2sources
+            if sendaction ~= "source-remove" then
+              -- If not a source remove, assume accept or add. Add the sources to participant2sources
+              for name, sourcelist in pairs(sources) do
+                module:log("debug", "Source add or accept, adding sources of type %s to participant2sources", name)
+                if participant2sources[room.jid][sender.nick][name] == nil then
+                  participant2sources[room.jid][sender.nick][name] = {}
+                end
+                for i, source in ipairs(sourcelist) do
+                  local sources_of_name = participant2sources[room.jid][sender.nick][name];
+                  local next_index = #sources_of_name + 1;
+                  module:log("debug", "Mapping source %s at index %d for participant %s", source.attr.ssrc, next_index, sender.nick)
+                  participant2sources[room.jid][sender.nick][name][next_index] = source
+                end
+              end
+            else -- Source remove
+              for name, sourcelist in pairs(sources) do
+                module:log("debug", "Source remove, removing sources of type %s from participant2sources", name)
+                -- Get the ssrcs we need to remove
+                local should_remove = {};
+                for i, source in ipairs(sourcelist) do
+                  module:log("debug", "Source remove source with ssrc %s from participant2sources", source.attr.ssrc)
+                  should_remove[source.attr.ssrc] = true
+                end
+
+                local sources_of_name = participant2sources[room.jid][sender.nick][name] or {}
+                for i=#sources_of_name,1,-1 do
+                  local source = participant2sources[room.jid][sender.nick][name][i]
+                  if should_remove[source.attr.ssrc] then
+                    table.remove(participant2sources[room.jid][sender.nick][name], i)
+                    module:log("debug", "Removed source with ssrc %s from participant2sources, new size: %d", source.attr.ssrc, #sources_of_name)
+                  end --End if should_remove
+                end --End iterating through array
```

Comment:
This feels like a potential resizing-array-while-iterating gotcha if there were multiple sources removed. Constructing a new array with the correct information and then swapping it in one step would be better.